backends/ref/ceed-ref-operator.c

```diff
@@ -435,6 +496,197 @@ static int CeedCompositeOperatorApply_Ref(CeedOperator op, CeedVector invec,
+        break;
+      case CEED_EVAL_WEIGHT:
+        break; // Not active input
+      case CEED_EVAL_DIV:
+        break; // Not implemented
+      case CEED_EVAL_CURL:
+        break; // Not implemented
+      }
+    }
+  }
+
+  // Count number of active output fields
+  for (CeedInt i=0; i<numoutputfields; i++) {
+    ierr = CeedQFunctionFieldGetEvalMode(qfoutputfields[i], &emode);
+    CeedChk(ierr);
+    // Get input vector
+    ierr = CeedOperatorFieldGetVector(opoutputfields[i], &vec); CeedChk(ierr);
+    // Check if active output
+    if (vec == CEED_VECTOR_ACTIVE) {
+      switch (emode) {
+      case CEED_EVAL_NONE:
+      case CEED_EVAL_INTERP:
+        ierr = CeedQFunctionFieldGetNumComponents(qfoutputfields[i], &ncomp);
+        CeedChk(ierr);
+        ierr = CeedVectorGetArray(impl->qvecsout[i], CEED_MEM_HOST, &tmp);
+        CeedChk(ierr);
+        ierr = CeedRealloc(numactiveout + ncomp, &activeout); CeedChk(ierr);
+        for (CeedInt field=0; field<ncomp; field++) {
+          activeout[numactiveout+field] = &tmp[field*Q];
+        }
+        numactiveout += ncomp;
+        ierr = CeedVectorRestoreArray(impl->qvecsout[i], &tmp); CeedChk(ierr);
+        break;
+      case CEED_EVAL_GRAD:
+        ierr = CeedOperatorFieldGetBasis(opoutputfields[i], &basis);
+        CeedChk(ierr);
+        ierr = CeedBasisGetDimension(basis, &dim); CeedChk(ierr);
+        ierr = CeedQFunctionFieldGetNumComponents(qfoutputfields[i], &ncomp);
+        CeedChk(ierr);
+        ierr = CeedVectorGetArray(impl->qvecsout[i], CEED_MEM_HOST, &tmp);
+        CeedChk(ierr);
+        ierr = CeedRealloc(numactiveout + ncomp*dim, &activeout); CeedChk(ierr);
+        for (CeedInt field=0; field<ncomp*dim; field++) {
+          activeout[numactiveout+field] = &tmp[field*Q];
+        }
+        numactiveout += ncomp*dim;
+        ierr = CeedVectorRestoreArray(impl->qvecsout[i], &tmp); CeedChk(ierr);
+        break;
+      case CEED_EVAL_WEIGHT:
+        break; // Not output
+      case CEED_EVAL_DIV:
+        break; // Not implemented
+      case CEED_EVAL_CURL:
+        break; // Not implemented
+      }
+    }
+  }
+
+  // Check sizes
+  if (!numactivein || !numactiveout) {
+    Ceed ceed;
+    ierr = CeedOperatorGetCeed(op, &ceed); CeedChk(ierr);
+    return CeedError(ceed, 1,
+                       "Cannot linearize qfunction without active inputs and outputs");
+  }
+
+  // Loop through elements
+  for (CeedInt e=0; e<numelements; e++) {
+    // Input basis apply
+    ierr = CeedOperatorInputBasis_Ref(e, Q, opinputfields, qfinputfields,
+                                      numinputfields, impl); CeedChk(ierr);
+
+    // Linearize QFunction
+    ierr = CeedQFunctionApply(qf, Q, impl->qvecsin, impl->qvecsout);
+    CeedChk(ierr);
+    // Copy A
+    for (CeedInt out=0; out<numactiveout; out++)
+      for (CeedInt q=0; q<Q; q++)
+        a[(e*numactiveout+out)*Q+q] = activeout[out][q];
+    // Compute B
+    CeedScalar epsilon = 1E-6, epsinv = 1E6;
```

Comment:
This is *a* finite differencing to linearize, but I'm not entirely happy with the loss of accuracy. I should look for a better way to decide upon epsilon.